### DISCUSSION TITLE
Does Directus validate hd parameter in returned ID token for OAuth2 SSO?

### DISCUSSION THREAD

**Discussion by chrisbartley** [directus/directus#11948] [Archive]
Just wondered whether someone could confirm that Directus does indeed validate the `hd` (hosted domain) parameter in the returned OAuth2 ID token?  I've looked in the code, but haven't found anything definitive yet.  But from what I can tell from some simple tests, it does, and that's magical and wonderful and yet another reason I freaking love Directus ♥️ 

More detail:  My users all have accounts within a particular Google Workspace.  So, they're Google accounts, but with a single, specific domain name in the email address.  I added SSO support to my Directus instance, following the instructions for [Google OpenID](https://docs.directus.io/configuration/sso/#google). Added the suggested params from step 7, with values specific to my use case, but also added this extra one so that the `hd` param gets sent along to [streamline the login process](https://developers.google.com/identity/protocols/oauth2/openid-connect#hd-param):

```JavaScript
AUTH_GOOGLE_PARAMS : {hd: 'mydomain.com'}
```

And that all works great.  Go to the app, there's now a **Log In with Google** button, and clicking it takes me directly to my company's login screen, and logging in bounces me right back to the Directus admin.  Amazing 👏 

The [docs for the `hd` param](https://developers.google.com/identity/protocols/oauth2/openid-connect#hd-param) warn that you shouldn't "rely on this UI optimization to control who can access your app, as client-side requests can be modified".  And so you should validate that the returned ID token has an `hd` claim that matches.  I tested by doing exactly that...manually modifying the request to remove the `hd` param from the query string, and logged into a Google account that's _not_ part of the company Google Workspace, and, when bounced back to Directus, I saw that Directus gave an error saying that the user was not authorized.  Excellent!

So, from all that, it appears the Directus is indeed validating the `hd` param, as specified in [step 5 of Google's docs for validating the ID token](https://developers.google.com/identity/protocols/oauth2/openid-connect#validatinganidtoken), yes? Apologies for my not being able to find it in the code...just hoping someone can provide confirmation.

Thanks! 🙏

**Comment by rijkvanzanten** [comment]
Heya! Apologies for the silence, this one took some more research..


We didn't add explicit checks for the `hd` parameter / response claims on the Directus side of the equation. We're relying on `node-openid` to make the requests to the third party for token checks. As far as I can tell from going over their codebase, it doesn't look like there's an explicit check against the `hd` claim in the tokens check, so I believe nothing in the chain is checking the validity of the `hd` claim returned by Google. 

I believe the reason that you're seeing the login fail (successfully) is because we rely on PKCE to prevent access when the authorization URL was modified outside of our link generation, but I'll have to tap in @aidenfoxx to confirm/deny that.

**Reply by chrisbartley** [reply]
Ah, gotcha, thank you for the extensive research into this! 🙏

**Reply by aidenfoxx** [reply]
Huh... This discussion slipped by me. But as Rijk says, `openid-client` doesn't verify the "hd" of the request. Not sure why the login failed when you manually changed the "hd" param in the URL, however. I would expect that to work, but that could just be because Google does a redirect to hide the "code_verifier" (which would cause PKCE to fail) or something? Not 100% sure.

Either way, I would not rely on "hd" as a protection. Ideally we'd have filters for dynamic user generation so you could limit the scope to pre-defined email domains.

There is a PR up for that. Just needs some love I think. https://github.com/directus/directus/pull/9363
